src/transformers/models/efficientloftr/modeling_efficientloftr.py

```diff
@@ -0,0 +1,1345 @@
+    )
+    return positional_embeddings
+
+
+class EfficientLoFTRAggregatedAttention(nn.Module):
+    def __init__(self, config: EfficientLoFTRConfig, layer_idx: int):
+        super().__init__()
+
+        self.q_aggregation_kernel_size = config.q_aggregation_kernel_size
+        self.aggregation = EfficientLoFTRAggregationLayer(config)
+        self.attention = EfficientLoFTRAttention(config, layer_idx)
+        self.mlp = EfficientLoFTRMLP(config)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        encoder_hidden_states: Optional[torch.Tensor] = None,
+        encoder_attention_mask: Optional[torch.Tensor] = None,
+        position_embeddings: Optional[tuple[torch.Tensor, torch.Tensor]] = None,
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> torch.Tensor:
+        batch_size, embed_dim, _, _ = hidden_states.shape
+
+        # Aggregate features
+        aggregated_hidden_states, attention_mask, encoder_hidden_states, encoder_attention_mask = self.aggregation(
+            hidden_states, attention_mask, encoder_hidden_states, encoder_attention_mask
+        )
+
+        attention_hidden_states = aggregated_hidden_states.reshape(batch_size, -1, embed_dim)
+        encoder_hidden_states = encoder_hidden_states.reshape(batch_size, -1, embed_dim)
+
+        if position_embeddings is not None:
+            position_embeddings = get_positional_embeddings_slice(aggregated_hidden_states, position_embeddings)
+            position_embeddings = tuple(tensor.reshape(batch_size, -1, embed_dim) for tensor in position_embeddings)
+
+        # Multi-head attention
+        message, _ = self.attention(
+            attention_hidden_states,
+            attention_mask,
+            encoder_hidden_states,
+            encoder_attention_mask,
+            position_embeddings=position_embeddings,
+            **kwargs,
+        )
+
+        # Upsample features
+        _, aggregated_h, aggregated_w, _ = aggregated_hidden_states.shape
+        # (batch_size, seq_len, embed_dim) -> (batch_size, embed_dim, h, w) with seq_len = h * w
+        message = message.permute(0, 2, 1)
+        message = message.reshape(batch_size, embed_dim, aggregated_h, aggregated_w)
+        if self.q_aggregation_kernel_size != 1:
+            message = torch.nn.functional.interpolate(
+                message, scale_factor=self.q_aggregation_kernel_size, mode="bilinear", align_corners=False
+            )
+        intermediate_states = torch.cat([hidden_states, message], dim=1)
+        intermediate_states = intermediate_states.permute(0, 2, 3, 1)
+        output_states = self.mlp(intermediate_states)
+        output_states = output_states.permute(0, 3, 1, 2)
+
+        hidden_states = hidden_states + output_states
+
+        return hidden_states
+
+
+class EfficientLoFTRLocalFeatureTransformerLayer(nn.Module):
```

Comment:
should inherit from gradient checkpointing!

Reply:
fixed in https://github.com/huggingface/transformers/pull/36355/commits/f59af0dfa22454158934b0dfcabd5f2339261bfc